runtime/runtime.c

```diff
@@ -329,4 +332,9 @@ void runtime_init() {
     if (runtime_heap_base == 0) {
         runtime_heap_base = memory_size;
     }
+
+    // The start component of a module declares the function index of a start function that is automatically invoked
+    // when the module is instantiated, after tables and memories have been initialized.
+    // https://webassembly.github.io/spec/core/syntax/modules.html#start-function
+    awsm_abi__start_fn();
```

Comment:
I have forgotten a bit about the initialization logic. Why isn't this just above where we currently call "main"?

Reply:
After messing around with wasm2c and seeing how Mao has been using our runtime, I now tend to think of `main` as outside of our runtime library. That sort of means that `runtime_init` is really something like `module_instantiate`. This basically is [the same pattern as the web](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/instantiate). The key thing is that the wasm `(start)` instruction is basically registering a function to run automatically during instantiation. 

So basically I'm treating `runtime_init` as the wasm instantiation function and placing the start invocation there per the spec: https://webassembly.github.io/spec/core/syntax/modules.html#start-function

This keeps our API clean for a consumer. For example, here's an example program and uses a library compiled as an aWsm module. This doesn't initialize WASI at all because this program doesn't use it.

```c
#include <stdbool.h>

#include "../../runtime/runtime.h"

IMPORT bool wasmf_isArmstrongNumber(int);

int main(int argc, char* argv[]) {
    runtime_init();

    awsm_assert(wasmf_isArmstrongNumber(0) == true);
    awsm_assert(wasmf_isArmstrongNumber(5) == true);
    awsm_assert(wasmf_isArmstrongNumber(10) == false);
    awsm_assert(wasmf_isArmstrongNumber(153) == true);
    awsm_assert(wasmf_isArmstrongNumber(100) == false);
    awsm_assert(wasmf_isArmstrongNumber(9474) == true);
    awsm_assert(wasmf_isArmstrongNumber(9475) == false);
    awsm_assert(wasmf_isArmstrongNumber(9926315) == true);
    awsm_assert(wasmf_isArmstrongNumber(9926314) == false);
}
```